File: src/com/train/algorithm/tree/implementInC++/UnivaluedBinaryTree.cpp
//
// Created by 罗斯 on 23/10/2022.
//

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool isUnivalTree(TreeNode* root) {
        if (root == nullptr) return true;

        if (root->left != nullptr && root->left->val != root->val) return false;
        if (root->right != nullptr && root->right->val != root->val) return false;

        return isUnivalTree(root->right) && isUnivalTree(root->left);
    }
};

File: src/com/train/algorithm/tree/implementInPython/UnivaluedBinaryTree.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution(object):
    def isUnivalTree(self, root):
        if not root:
            return True
        if root.right:
            if root.val != root.right.val:
                return False
        if root.left:
            if root.val != root.left.val:
                return False

        return self.isUnivalTree(root.right) and self.isUnivalTree(root.left)